### PR TITLE
Catches crashes and logs to Crashlytics

### PR DESCRIPTION
**Pull request by dtvcfinc**
Catches the crashes reported in crashlytics and logs more information like process id, process name, thread id, etc

### CODE REVIEW

File: src/main/java/com/sukesan1984/stepsensorlib/Database.java [hunk 1]
```diff
@@ -92,23 +103,45 @@ public Cursor query(final String[] columns, final String selection,
...
-            db.setTransactionSuccessful();
-        } finally {
-            db.endTransaction();
+            SQLiteDatabase db = getWritableDatabase();
+            db.beginTransaction();
+            try {
+                updateOrInsertWithoutTransaction(db, dateAndHour, stepsSinceBoot);
+                db.setTransactionSuccessful();
+            } finally {
+                db.endTransaction();
```
Comment (ypresto) [inline src/main/java/com/sukesan1984/stepsensorlib/Database.java]: Missing db.close()

File: src/main/java/com/sukesan1984/stepsensorlib/Database.java [hunk 2]
```diff
@@ -92,23 +103,45 @@ public Cursor query(final String[] columns, final String selection,
...
+            }
+        } catch (SQLiteDatabaseLockedException | IllegalStateException e) {
+            Crashlytics.logException(e);
         }
     }
 
     private void updateOrInsertWithoutTransaction(SQLiteDatabase db, long dateAndHour, int stepsSinceBoot) {
         Cursor c = getByDateAndHour(dateAndHour);
```
Comment (ypresto) [inline src/main/java/com/sukesan1984/stepsensorlib/Database.java]: Missing `c.close()`

File: src/main/java/com/sukesan1984/stepsensorlib/Database.java [hunk 3]
```diff
@@ -135,6 +168,10 @@ private void updateOrInsertWithoutTransaction(SQLiteDatabase db, long dateAndHou
     public void insertOrReplaceSteps(final long dateAndHour, final int steps) {
         SQLiteDatabase db = getWritableDatabase();
         Cursor c = getByDateAndHour(dateAndHour);
```
Comment (ypresto) [inline src/main/java/com/sukesan1984/stepsensorlib/Database.java]: Missing matching `c.close()`

File: src/main/java/com/sukesan1984/stepsensorlib/Database.java [hunk 4]
```diff
@@ -321,12 +367,17 @@ public int getTodayStep() {
...
     @NonNull
     public List<ChunkStepCount> getNotRecordedChunkedStepCounts() {
-        Cursor c = getReadableDatabase()
-                .query(TABLE_NAME, new String[]{COLUMN_DATE_AND_HOUR, COLUMN_STEPS},
-                        COLUMN_DATE_AND_HOUR + " != ? and " +
-                                COLUMN_IS_RECORDED_ON_SERVER + " = ?", new String[]{"-1", "0"}, null, null, null);
-        Logger.log("Not recoreded Chunk Size: " + c.getCount());
-        return createChunkedStepCounts(c);
+        try {
+            Cursor c = getReadableDatabase()
```
Comment (ypresto) [inline src/main/java/com/sukesan1984/stepsensorlib/Database.java]: Missing matching `c.close()`

**Comment by ypresto** [comment]
Cursor leak can block next transaction, I think, while I didn't tested it.

**Comment by ypresto** [comment]
I tested but the issue does not reproduced with transaction while cursor opened.

**Comment by dtvcfinc** [comment]
@ypresto I tested with thousand threads with random delays and I wasn't able to reproduce it, I don't think it is related with the cursors, most probably it is a different process that's why logging this could be useful to know the real cause

**Comment by sukesan1984** [comment]
please add c.close!
thanks.

File: src/main/java/com/sukesan1984/stepsensorlib/Database.java [hunk 5]
```diff
@@ -196,6 +197,7 @@ public void insertOrReplaceSteps(final long dateAndHour, final int steps) {
         } finally {
             db.endTransaction();
         }
+        c.close();
```
Comment (ypresto) [inline src/main/java/com/sukesan1984/stepsensorlib/Database.java]: It should be placed in `finally` block.

Comment (sukesan1984) [inline src/main/java/com/sukesan1984/stepsensorlib/Database.java]: oh... sorry @dtvcfinc 
my fault.

**Pull request merged**